File: src/features/typing/data/go/sliding_window_maximum.go
package main

func maxSlidingWindow(nums []int, k int) []int {
	deque := []int{}
	result := []int{}
	for i, num := range nums {
		for len(deque) > 0 && nums[deque[len(deque)-1]] <= num {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, i)
		if deque[0] <= i-k {
			deque = deque[1:]
		}
		if i >= k-1 {
			result = append(result, nums[deque[0]])
		}
	}
	return result
}

File: src/features/typing/data/dart/sliding_window_maximum.dart
class Solution {
  List<int> maxSlidingWindow(List<int> nums, int k) {
    final deque = <int>[];
    final result = <int>[];
    for (int i = 0; i < nums.length; i++) {
      while (deque.isNotEmpty && nums[deque.last] <= nums[i]) {
        deque.removeLast();
      }
      deque.add(i);
      if (deque.first <= i - k) {
        deque.removeAt(0);
      }
      if (i >= k - 1) {
        result.add(nums[deque.first]);
      }
    }
    return result;
  }
}
